### ISSUE TITLE
`Matrix(Integer,[[1]])` returns a `IsIntegerMatrix` but does not implement the `MatrixObj` protocol

### ISSUE THREAD

**Issue by fingolfin** [semigroups/Semigroups#807]
The following happens with both Semigroups 3.4.2 and 4.0.0:

Without it:
```gap
gap> Matrix(Integers, [[1,2],[3,4]]);
<2x2-matrix over Integers>
gap> BaseDomain(last);
Integers
```
With semigroups loaded:
```
gap> Matrix(Integers, [[1,2],[3,4]]);
Matrix(IsIntegerMatrix, [[1, 2], [3, 4]])
gap> BaseDomain(last);
Error, no method found!
```

There are two options to resolve this:
1. implement `BaseDomain` and possibly other methods needed to make `IsIntegerMatrix` compliant with the `MatrixObj` interface (to the best of  your ability -- we definitely need to improve the docs, and add conformance tests)
2. don't do this, but then also don't change the behaviour of the `Matrix(Integers, M)` constructor

Observed in https://github.com/gap-system/gap/pull/4805

**Comment by wilfwilson** [comment]
This is related to #512.

**Comment by wilfwilson** [comment]
And I did some unfinished work in this direction in #621.

**Comment by fingolfin** [comment]
In this particular case, though, just adding a `BaseDomain` method for `IsIntegerMatrix` might already help, and shouldn't be hard or cause problems for you, should it? (Granted, I haven't looked at the code at all, so maybe I am mistaken here).

**Comment by james-d-mitchell** [comment]
I’m happy enough to try to resolve this, but would like to point out that I think this is an issue caused by changes in GAP rather than any change in Semigroups. In particular If I recall the code for creating integer matrices using ‘Matrix’ in Semigroups predates the code in GAP. In some sense this is the GAP library changing the behaviour of Semigroups rather than the other way around. This is just for the record, it’s undoubtedly better to have less code for matrices in Semigroups and more in the GAP library, and I’m happy to make whatever changes might be required.

**Comment by fingolfin** [comment]
> I’m happy enough to try to resolve this

Thanks!

> [...], but would like to point out that I think this is an issue caused by changes in GAP rather than any change in Semigroups.

You are 100% right, and I misrepresented this above (not intentionally, but still); in fact earlier today I spent digging through the issue & PR @wilfwilson linked to, re-reading things that I'd forgotten about (including my own comments 😅). It's an annoyingly complex and far too long-unresolved storey 😞 

> In particular If I recall the code for creating integer matrices using ‘Matrix’ in Semigroups predates the code in GAP. In some sense this is the GAP library changing the behaviour of Semigroups rather than the other way around.

Indeed! IMHO the real problem here is that GAP doesn't have namespaces or some other real strategy for encapsulation of modules, and hence clashes (esp. for nice & sweet names like `Matrix`) are difficult to avoid.

> This is just for the record, it’s undoubtedly better to have less code for matrices in Semigroups and more in the GAP library, and I’m happy to make whatever changes might be required.

Thank you!

**Comment by james-d-mitchell** [comment]
This is related to:

https://github.com/gap-system/gap/issues/4814

Basically, the package (against the `master` branch of GAP) is completely broken by the changes detailed in:

[Recent change to `Matrix` causes `Semigroups` package CI to break](https://github.com/gap-system/gap/issues/4814)

**Comment by james-d-mitchell** [comment]
> IMHO the real problem here is that GAP doesn't have namespaces or some other real strategy for encapsulation of modules, and hence clashes (esp. for nice & sweet names like Matrix) are difficult to avoid.

I agree, but we are where we are!

Reading over @wilfwilson's comments in PR #621, I think the "best" thing to do would be to somehow convert the integer and finite field matrices in `Semigroups` to be `MatrixObj`s while leaving matrices over other semirings as they are. 

I think fixing this for integer matrices might be more or less straightforward, we can just start using GAP's integer matrices in `Semigroups`. Maybe it won't be that easy, but I can't think of any obstacles right away.

I'm not so sure about matrices over finite fields, it seems that GAP does not allow 0-dimensional matrices over finite fields, and 0-dimensional matrices are crucial for computing semigroups of matrices over finite fields (I don't exactly remember the details but this somehow comes up and was a major point of pain when Markus and I implemented the stuff in Semigroups for semigroups of matrices over finite fields). We could replace all calls to `Matrix(GF(2), something)` by something else. But this would be backwards incompatible, and I'm not sure I want to do this. I don't really like the idea that our (5 year old) working code can be broken by an upstream change in GAP, and that the onus is on us to make changes to resolve the issue. 

I'm game to fix this, but I don't really know how to get started, don't want to break backwards compatibility unless it's absolutely necessary. If anyone has any ideas, I'll be happy to hear them.

**Comment by wilfwilson** [comment]
> Reading over @wilfwilson's comments in PR #621, I think the "best" thing to do would be to somehow convert the integer and finite field matrices in `Semigroups` to be `MatrixObjs` while leaving matrices over other semirings as they are.

@james-d-mitchell You might want to take a look at the diff on PR #621 (especially the main `EXPERIMENT:` commit, which does all the work). The PR goes some way to achieving exactly this. I'll post a comment there with more details.

**Comment by james-d-mitchell** [comment]
I believe that this is resolved in v4.0.1, if so I'll close the issue, thoughts @fingolfin and @wilfwilson ?

**Comment by fingolfin** [comment]
I think PR #621 is needed to resolve this

**Comment by james-d-mitchell** [comment]
My current (short term) plan for this is to do the following:

1.  Remove `IsIntegerMatrix` from Semigroups completely and instead use the integer matrices from the library. I have some partially working (incomplete) code that does this, and it seems to be more or less straightforward, only requiring changes in the Semigroups package. This will avoid having duplicate code in Semigroups and the main GAP library, reducing confusion, and minimize future problems 
2. Remove as much of `IsMatrixOverFiniteField` from Semigroups as possible. This will be more complicated than 1, and it might require making some changes to GAP itself (or maybe removing some of the features in Semigroups to the GAP library). I don't know how feasible this is, it might be more or less straightforward it might be super complicated, we'll have to wait and see. 
3. The other matrices in Semigroups will remain exactly as they are now (at least for the time being). I think there's no point in converting the other Semigroups matrices to fulfil the MatrixObj interface, at least there's no compelling reason to do it now, and the not making changes is quicker than making them.

Does this make sense? @fingolfin @wilfwilson @ahulpke

**Comment by james-d-mitchell** [comment]
This is related to https://github.com/gap-system/gap/issues/4878

**Comment by hulpke** [comment]
Thank you -- yes this makes sense. If there is any feature the integer (or finite field) matrices in the library are lacking, please let me know.

**Comment by james-d-mitchell** [comment]
It seems that everything works for integer matrices without too much pain. There is still the issue that it is difficult to recognise matrix objs and collections of them, that they don't belong to `IsAssociativeElement` or even `IsMultiplicativeElement`, but as the passing tests against GAP `master` indicate it's possible to work around all of these. I'll try matrices over finite fields next, and keep you up to date.

**Comment by james-d-mitchell** [comment]
@hulpke and @fingolfin I've removed the code for matrices over finite fields and integers in Semigroups in PR #827, and this seems to work (at least on my computer) with the current GAP master branch. It is not backwards compatible with GAP 4.10 or 4.11, and I'm not 100% sure what to do. 

I'll be happy to make a release of Semigroups that requires GAP 4.12 and includes the changes in PR #827, but I'm not sure if this is desirable before GAP 4.12 is released? Any thoughts?

**Comment by fingolfin** [comment]
@james-d-mitchell thank you for working this, and sorry for me not working on this, but I've been busy with other things (such as migrating the GAP infrastructure... which *finally* was done yesterday. I am not saying *completed* because more work is needed...). 

Anyway, getting the clash between GAP 4.12 and semigroups released is from POV the *one* blocker for a release of GAP 4.12, which is long overdue, so it is really important to get this resolved. I'll try to focus on it again and help.

Your PR removes a lot of code from Semigroups, but I am now wondering if this doesn't affect you or your users negatively? "Punishing" semigroups due to bad choices we may have made in the past is not something I am happy about :-(. That said, I can't evaluate at all what the impact of your PR #827 really is for you... if you say it is acceptable, and the "only" serious problem is to coordinate this with the release of GAP 4.12, we could overcome this (there certainly are more problems, where the GAP MatrixObj code needs to be improved, but the question is whether it needs to be in GAP 4.12.0 or could also come in 4.12.1 or 4.13.0 -- BTW, the GAP Days in October once again are also about MatrixObj and I hope to use substantially parts of it to further this agenda).

Perhaps we can have a videocall about this next week, to hash out a plan, assuming you have time? Unfortunately I have to run now for family obligations

**Comment by james-d-mitchell** [comment]
Thanks @fingolfin, maybe I chose my words poorly, I removed the code for matrices over finite fields that overlapped and interacted badly with MatrixObj from Semigroups, and I also made the existing functionality for semigroups and monoids of matrices over finite fields work, which I think is the optimal solution. So, there's no punishment (other than I had to get the code to work, but this was also not that difficult once I figured out the correct approach). 

I thought that this might be a blocker for GAP 4.12, which is why I wanted to resolve it, I'm also sorry that I wasn't able to earlier. 

Happy to talk next week and hash out a plan, let's be in touch via email to coordinate.
